sw/host/ot_certs/src/x509/tests.rs

```diff
@@ -0,0 +1,571 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use base64::{engine::general_purpose, Engine as _};
+use num_bigint_dig::BigUint;
+use std::collections::HashMap;
+use std::iter::zip;
+
+use foreign_types::{ForeignType, ForeignTypeRef};
+use openssl::asn1::{Asn1IntegerRef, Asn1Object, Asn1OctetStringRef, Asn1StringRef, Asn1Time};
+use openssl::ec::EcKey;
+use openssl::pkey::Public;
+use openssl::x509::{X509NameRef, X509};
+
+use crate::offsets::CertificateWithVariables;
+use crate::template::*;
+use crate::types::Type;
+use crate::x509::extension::*;
+use crate::x509::*;
+
+// Unfortunately, the rust openssl binding does not have an API to extract arbitrary extension but it exports
+// the low level function from the C library to do that.
+fn x509_get_ext_by_obj<'a>(x509: &'a X509, obj: &Asn1Object) -> Result<&'a Asn1OctetStringRef> {
+    unsafe {
+        // SAFETY the rust openssl binding guarantees that x509 and obj are valid objects.
+        let index = openssl_sys::X509_get_ext_by_OBJ(x509.as_ptr(), obj.as_ptr(), -1);
```

Comment:
Should we add a binding to the openssl crate upstream?

Reply:
You mean for the whole function? We could try. The last time I checked, the maintainer of the crate was sure how to handle custom extensions on the rust side, but this would hopefully be be noncontroversial.